### PR TITLE
soundwire: cadence_master: set frame shape and divider based on actual clk freq

### PR DESCRIPTION
**Pull request by bardliao**
"soundwire: cadence_master: set frame shape and divider based on actual clk freq" is the only new commit. CI reports below error without this commit.
```
[  405.383015] soundwire sdw-master-0-0: Controller Timed out on bank switch
[  405.383126] soundwire sdw-master-0-0: multi link bank switch failed: -110
```
The issue happens randomly when cdns_init_clock_ctrl() and cdns_bus_conf() set different divider. In theory, divider could be changed. But, the existing code set max_clk_freq in cdns_init_clock_ctrl() and set curr_dr_freq in sdw_initialize_slave() which doesn't make sense to me. The PR suggests setting curr_dr_freq if it is present.

To reproduce the issue:
Run 2 times for each below command one by one on LNL_SDW_AIOC devices.
TPLG=/lib/firmware/intel/sof-ipc4-tplg/sof-lnl-rt711-l0-rt1316-l23-rt714-l1.tplg MODEL=LNLM_SDW_AIOC SOF_TEST_INTERVAL=5 ~/sof-test/test-case/check-playback.sh -d 3 -l 3 -r 1
TPLG=/lib/firmware/intel/sof-ipc4-tplg/sof-lnl-rt711-l0-rt1316-l23-rt714-l1.tplg MODEL=LNLM_SDW_AIOC SOF_TEST_INTERVAL=5 ~/sof-test/test-case/check-capture.sh -d 3 -l 3 -r 1
TPLG=/lib/firmware/intel/sof-ipc4-tplg/sof-lnl-rt711-l0-rt1316-l23-rt714-l1.tplg MODEL=LNLM_SDW_AIOC SOF_TEST_INTERVAL=5 ~/sof-test/test-case/check-playback.sh -d 3 -l 1 -r 3
TPLG=/lib/firmware/intel/sof-ipc4-tplg/sof-lnl-rt711-l0-rt1316-l23-rt714-l1.tplg MODEL=LNLM_SDW_AIOC SOF_TEST_INTERVAL=5 ~/sof-test/test-case/check-capture.sh -d 3 -l 1 -r 3

### CODE REVIEW

**Comment by plbossart** [comment]
not following why you have 2 PRs that look mostly identical @bardliao? Which one do you want us to review?

**Comment by bardliao** [comment]
> not following why you have 2 PRs that look mostly identical @bardliao? Which one do you want us to review?

Please review #5160. This PR is to discuss the https://github.com/thesofproject/linux/pull/5179/commits/60f5a6b950c6e3bd6965c32a3c68653e66180e5b commit only

**Comment by bardliao** [comment]
SOFCI TEST

**Comment by bardliao** [comment]
Rebase to the latest topic/sof-dev branch and the CI test result looks good to me.

File: drivers/soundwire/cadence_master.c [hunk 1]
```diff
@@ -1360,13 +1361,25 @@ static void cdns_init_clock_ctrl(struct sdw_cdns *cdns)
...
 	cdns_updatel(cdns, CDNS_MCP_CLK_CTRL0,
 		     CDNS_MCP_CLK_MCLKD_MASK, divider);
 	cdns_updatel(cdns, CDNS_MCP_CLK_CTRL1,
 		     CDNS_MCP_CLK_MCLKD_MASK, divider);
 
+	/* Set frame shape base on the actual bus frequency. */
+	if (!prop->default_frame_rate || !prop->default_row) {
+		dev_err(cdns->dev, "Default frame_rate %d or row %d is invalid\n",
+			prop->default_frame_rate, prop->default_row);
+		return;
```
Comment (ranj063) [inline drivers/soundwire/cadence_master.c]: I think we should return an error in this case, so change the return type from void to int for this function?

Comment (bardliao) [reply drivers/soundwire/cadence_master.c]: > I think we should return an error in this case, so change the return type from void to int for this function?

Agree, done

File: drivers/soundwire/cadence_master.c [hunk 2]
```diff
@@ -1345,13 +1345,14 @@ static u32 cdns_set_initial_frame_shape(int n_rows, int n_cols)
 	return val;
 }
 
-static void cdns_init_clock_ctrl(struct sdw_cdns *cdns)
```
Comment (ujfalusi) [inline drivers/soundwire/cadence_master.c:1348]: typo in commit message: `Frame shap` and drop the leading `And` from the second sentence.

**Review by ranj063 — APPROVED** [review]

**Review by ujfalusi — APPROVED** [review]

**Pull request merged**